### PR TITLE
[test-only] fix flaky operator tests

### PR DESCRIPTION
**Pull request by richardwxn**
fix the issue in postsubmit test

### CODE REVIEW

File: pilot/pkg/serviceregistry/kube/controller/namespacecontroller.go [hunk 1]
```diff
@@ -142,8 +142,10 @@ func (nc *NamespaceController) insertDataForNamespace(ns string) error {
 func (nc *NamespaceController) namespaceChange(ns *v1.Namespace) error {
 	if ns.Status.Phase != v1.NamespaceTerminating {
 		return nc.insertDataForNamespace(ns.Name)
+	} else {
+		fmt.Printf("namespace: %s terminating, not updating configmap", ns.Name)
+		return nc.insertDataForNamespace(ns.Name)
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/namespacecontroller.go:147]: Could not understand even with the log

File: istioctl/cmd/workload.go [hunk 2]
```diff
@@ -364,7 +366,17 @@ func createMeshConfig(kubeClient kube.ExtendedClient, wg *clientv1alpha3.Workloa
...
+	istio := &v1.ConfigMap{}
+	err := retry.UntilSuccess(func() error{
+		var err error
+		istio, err = kubeClient.CoreV1().ConfigMaps(istioNamespace).Get(context.Background(), istioCM, metav1.GetOptions{})
+		if err != nil {
+			fmt.Println("configmap istio-ca-root-cert not found, retry waiting")
+			return fmt.Errorf("configmap %s was not found in namespace %s: %v", controller.CACertNamespaceConfigMap, wg.Namespace, err)
+		}
+		return nil
+	}, retry.Timeout(5*time.Minute))
```
Comment (esnible) [inline istioctl/cmd/workload.go:378]: Would it make sense for `istioctl x workload entry configure` to have a `--timeout duration` flag similar to `istioctl x wait`?

**Comment by richardwxn** [comment]
@howardjohn can you PTAL?

File: tests/integration/operator/switch_cr_test.go [hunk 3]
```diff
@@ -115,61 +117,50 @@ func TestController(t *testing.T) {
...
+					}
 				}
 				return nil
 			}, retry.Timeout(retryTimeOut), retry.Delay(retryDelay))
 
-			// cleanup created resources
 			t.Cleanup(func() {
 				scopes.Framework.Infof("cleaning up resources")
+				unInstallCmd := []string{
+					"x", "uninstall", "--purge", "--skip-confirmation",
```
Comment (howardjohn) [inline tests/integration/operator/switch_cr_test.go:143]: this doesn't match https://istio.io/latest/docs/setup/install/operator/#uninstall, should the doc be updated?

File: tests/integration/operator/switch_cr_test.go [hunk 4]
```diff
@@ -115,61 +117,50 @@ func TestController(t *testing.T) {
...
+					}
+					if dp, _ := cs.AppsV1().Deployments(OperatorNamespace).Get(context.TODO(), n, kubeApiMeta.GetOptions{}); dp.Name != "" {
+						return fmt.Errorf("got operator deployment %s from cluster, expected to be removed", dp.Name)
+					}
 				}
 				return nil
 			}, retry.Timeout(retryTimeOut), retry.Delay(retryDelay))
 
-			// cleanup created resources
 			t.Cleanup(func() {
```
Comment (howardjohn) [inline tests/integration/operator/switch_cr_test.go:140]: not sure t.Cleanup() here is correct? t.cleanup means after the test is complete, we will *always* run this, even if previous steps failed.

So if we want to always run it, we need to call it earlier, and if not we don't need to wrap it in t.Cleanup.

I do think we probably want to use t.Cleanup, otherwise if the test fails on earlier steps it will not cleanup and all future tests will also fail.

File: manifests/charts/istio-operator/templates/clusterrole.yaml [hunk 5]
```diff
@@ -91,6 +91,14 @@ rules:
   - rolebindings
   verbs:
   - '*'
+- apiGroups:
```
Comment (howardjohn) [inline manifests/charts/istio-operator/templates/clusterrole.yaml:94]: Why is this needed? is the operator broken currently?

Comment (richardwxn) [reply manifests/charts/istio-operator/templates/clusterrole.yaml:94]: yes, it failed because it did not have permission to get the lease

File: tests/integration/operator/switch_cr_test.go (hunk 3 above)
Comment (richardwxn) [reply tests/integration/operator/switch_cr_test.go:143]: this corresponds to this section instead: https://istio.io/latest/docs/setup/install/istioctl/#uninstall-istio

File: tests/integration/operator/switch_cr_test.go [hunk 6]
```diff
@@ -77,6 +79,9 @@ func TestController(t *testing.T) {
 				t.Fatal(err)
 			}
 			cleanupInClusterCRs(t, cs)
+			t.Cleanup(func() {
+				cleanupIstioResources(t, cs, istioCtl)
+			})
```
Comment (morvencao) [inline tests/integration/operator/switch_cr_test.go:84]: Is it necessary to manually cleanup the istio resources after all the istio operator CRs are deleted, the in-cluster operator will cleanup the istio resources. right?

**Review by esnible — APPROVED** [review]

File: tests/integration/operator/switch_cr_test.go (hunk 6 above)
Comment (richardwxn) [reply tests/integration/operator/switch_cr_test.go:84]: @morvencao makes sense, we should probably check whether the istio resources are cleaned up as part of test after deleting the CR. But anyways there are still some other resources that require manually clean up, for example the secrets/configmaps created dynamically by istiod.

File: tests/integration/operator/switch_cr_test.go (hunk 4 above)
Comment (richardwxn) [reply tests/integration/operator/switch_cr_test.go]: move t.cleanUp to correct position

**Review by morvencao — APPROVED** [review]

**Review by elfinhe — APPROVED** [review]

**Comment by howardjohn** [comment]
Looks good, but since the test just failure its probably still flaky?

**Comment by richardwxn** [comment]
@howardjohn the most recent failure was not related to previous flakiness: the cleanup problem of hanging configmaps but other reason and I fixed that.

**Review by howardjohn — APPROVED** [review]
